### PR TITLE
EICNET-1858: Group ownership transfer confirmation improvements

### PR DESCRIPTION
**Pull request by 4hardfire**
### Improvements

- Change request timeout field to a select list with options. See [EICNET-1931](https://citnet.tech.ec.europa.eu/CITnet/jira/browse/EICNET-1931)
- Add request expiration date in message notification. See [EICNET-1932](https://citnet.tech.ec.europa.eu/CITnet/jira/browse/EICNET-1932)
- Show request expiration date in group admin members overview. See [EICNET-1933](https://citnet.tech.ec.europa.eu/CITnet/jira/browse/EICNET-1933)
- Provide feature to allow GO/DA/SA/SCM users to cancel requests to transfer ownership. See [EICNET-1930](https://citnet.tech.ec.europa.eu/CITnet/jira/browse/EICNET-1930)

### Tests

Improvements:

- [x] As GO, go to a group members page `/groups/<group-name>/members`

EICNET-1931:

- [x] Request ownership transfer to a GA and make sure the timeout field is a select list
- [x] Click on request transfer ownership to create the request

EICNET-1933 (**DO NOT** test EICNET-1932 before this one):

- [x] Go back to the group members page and check if the expiration date is shown in the list bellow the GA that got requested

EICNET-1932:

- [x] As SA, run cron
- [x] Check if the new GA receives a notification and the expiration date is presented

EICNET-1930 - Cancel request ownership transfer:

- [x] As GO, go to a group members page `/groups/<group-name>/members`
- [x] Request ownership transfer to a GA
- [x] Go back to the group members page and make sure you see a link to cancel the request in the list bellow the GA that got requested
- [x] Click on "Cancel ownership transfer" and provide a reason why you want to cancel the request and confirm
- [x] As SA, run cron
- [x] Check if the requested user received a notification

### CODE REVIEW

File: lib/modules/eic_flags/src/Service/AbstractRequestHandler.php [hunk 1]
```diff
@@ -429,6 +440,18 @@ public function canCloseRequest(
...
 
+  /**
+   * {@inheritdoc}
+   */
+  public function canCancelRequest(
+    AccountInterface $account,
+    ContentEntityInterface $entity
+  ) {
+    // Default access.
+    $access = AccessResult::forbidden();
```
[the remark this answers is not in the document]
Comment (lramarojaona) [reply lib/modules/eic_flags/src/Service/AbstractRequestHandler.php:451]: Ok, sorry, this is an abstract class

File: lib/modules/eic_flags/src/Service/TransferOwnershipRequestHandler.php [hunk 2]
```diff
@@ -353,12 +389,126 @@ private function canCloseRequestGroupTransferOwnership(
...
+    }
+
+    // We return access denied if there are no requests for this entity.
+    if (!$this->hasOpenRequest($entity, $account)) {
+      return $access;
+    }
+
+    $requests = $this->getOpenRequests($entity);
+    $request = reset($requests);
+    $expiration_date = 0;
```
Comment (lramarojaona) [inline lib/modules/eic_flags/src/Service/TransferOwnershipRequestHandler.php:466]: Maybe add a comment that the feature is designed to have one open request at a time only?

**Review by lramarojaona — CHANGES_REQUESTED** [review]

File: lib/modules/eic_flags/src/Form/RequestCancelForm.php [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    $handler = $this->collector->getHandlerByType($this->requestType);
+    if (!$handler instanceof HandlerInterface) {
+      throw new InvalidArgumentException('Request type is invalid');
+    }
+
+    $flag_id = $handler->getFlagId($this->entity->getEntityTypeId());
+    /** @var \Drupal\flag\FlaggingInterface $flag */
+    $flag = $this->flagService->getFlagById($flag_id);
+    if (!$flag instanceof FlagInterface) {
+      throw new InvalidArgumentException('Flag doesn\'t exists');
```
[the remark this answers is not in the document]
Comment (4hardfire) [reply lib/modules/eic_flags/src/Form/RequestCancelForm.php]: Done!

File: lib/modules/eic_flags/src/Service/TransferOwnershipRequestHandler.php (hunk 2 above)
Comment (4hardfire) [reply lib/modules/eic_flags/src/Service/TransferOwnershipRequestHandler.php:468]: Done!

File: lib/modules/eic_flags/src/Form/RequestCancelForm.php [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity.repository'),
+      $container->get('entity_type.bundle.info'),
+      $container->get('datetime.time'),
```
[the remark this answers is not in the document]
Comment (4hardfire) [reply lib/modules/eic_flags/src/Form/RequestCancelForm.php]: Done!

**Review by lramarojaona — APPROVED** [review]

**Pull request merged**